query/query/query.py

```diff
@@ -52,107 +66,97 @@ async def healthcheck(request):  # pylint: disable=unused-argument
 
 
-def blocking_execute(jbackend, username, session_id, billing_project, bucket, code):
-    return jbackend.execute(username, session_id, billing_project, bucket, code)
+def blocking_execute(jbackend, userdata, body):
+    return jbackend.execute(userdata['username'], userdata['session_id'], body['billing_project'], body['bucket'], body['code'])
 
 
-@routes.post('/execute')
-@rest_authenticated_users_only
-async def execute(request, userdata):
-    app = request.app
-    thread_pool = app['thread_pool']
-    jbackend = app['jbackend']
-    body = await request.json()
-    billing_project = body['billing_project']
-    bucket = body['bucket']
-    code = body['code']
-    log.info(f'execute: {code}')
-    await add_user(app, userdata)
-    jresp = await blocking_to_async(thread_pool, blocking_execute, jbackend, userdata['username'], userdata['session_id'], billing_project, bucket, code)
-    return java_to_web_response(jresp)
+def blocking_value_type(jbackend, userdata, body):
+    return jbackend.valueType(userdata['username'], body['code'])
 
 
-def blocking_value_type(jbackend, username, code):
-    return jbackend.valueType(username, code)
+def blocking_table_type(jbackend, userdata, body):
+    return jbackend.tableType(userdata['username'], body['code'])
 
 
-@routes.post('/type/value')
-@rest_authenticated_users_only
-async def value_type(request, userdata):
-    app = request.app
-    thread_pool = app['thread_pool']
-    jbackend = app['jbackend']
-    code = await request.json()
-    log.info(f'value type: {code}')
-    await add_user(app, userdata)
-    jresp = await blocking_to_async(thread_pool, blocking_value_type, jbackend, userdata['username'], code)
-    return java_to_web_response(jresp)
+def blocking_matrix_type(jbackend, userdata, body):
+    return jbackend.matrixTableType(userdata['username'], body['code'])
 
 
-def blocking_table_type(jbackend, username, code):
-    return jbackend.tableType(username, code)
+def blocking_blockmatrix_type(jbackend, userdata, body):
+    return jbackend.blockMatrixType(userdata['username'], body['code'])
 
 
-@routes.post('/type/table')
-@rest_authenticated_users_only
-async def table_type(request, userdata):
+def blocking_get_reference(jbackend, userdata, body):   # pylint: disable=unused-argument
+    return jbackend.referenceGenome(userdata['username'], body['name'])
+
+
+async def handle_ws_response(request, userdata, cmd_str, f):
+    log.info('connecting websocket')
     app = request.app
-    thread_pool = app['thread_pool']
     jbackend = app['jbackend']
-    code = await request.json()
-    log.info(f'table type: {code}')
+
     await add_user(app, userdata)
-    jresp = await blocking_to_async(thread_pool, blocking_table_type, jbackend, userdata['username'], code)
-    return java_to_web_response(jresp)
+    log.info('connecting websocket')
+    ws = web.WebSocketResponse(heartbeat=30, max_msg_size=0)
+    task = None
+    try:
+        await ws.prepare(request)
```

Comment:
prepare needs to happen outside the `try`, if anything goes wrong before `_post_start` is called in `prepare`, the `_writer` will be `None` and `close` will raise an exception.